include/dali/imgcodec/image_decoder.h

```diff
@@ -0,0 +1,366 @@
+                                   cspan<ImageSource *> in,
+                                   DecodeParams opts,
+                                   cspan<ROI> rois = {});
+
+  std::vector<DecodeResult> Decode(DecodeContext ctx,
+                                   TensorList<GPUBackend> &out,
+                                   cspan<ImageSource *> in,
+                                   DecodeParams opts,
+                                   cspan<ROI> rois = {});
+  /**
+   * @brief Sets a value of a parameter.
+   *
+   * It sets a value of a parameter for all existing sub-decoders as well ones that will be
+   * created in the future.
+   *
+   * This function succeeds even if no sub-decoder recognizes the key.
+   * If the value is incorrect for one of the decoders, but that decoder has not yet
+   * been constructed, an error may be thrown at a later time, when the decoder is instantiated.
+   */
+  bool SetParam(const char *key, const any &value) override;
+
+  int SetParams(const std::map<std::string, any> &params) override;
+
+  /**
+   * @brief Gets a value previously passed to `SetParam` with the given key.
+   */
+  any GetParam(const char *key) const override;
+
+ private:
+  /**
+   * @brief Describes a sub-batch of work to be processed
+   *
+   * This object contains a (shared) promise from the original request containing the full batch
+   * and a mapping of indices from this sub-batch to the full batch.
+   * It also contains a subset of relevant sample views, sources, etc.
+   */
+  struct ScheduledWork {
+    ScheduledWork(DecodeContext ctx, DecodeResultsPromise results)
+    : ctx(std::move(ctx)), results(std::move(results)) {}
+
+    void clear() {
+      indices.clear();
+      sources.clear();
+      cpu_outputs.clear();
+      gpu_outputs.clear();
+      rois.clear();
+    }
+
+    int num_samples() const {
+      return indices.size();
+    }
+
+    bool empty() const {
+      return indices.empty();
+    }
+
+    void resize(int num_samples) {
+      indices.resize(num_samples);
+      sources.resize(num_samples);
+      if (!cpu_outputs.empty())
+        cpu_outputs.resize(num_samples);
+      if (!gpu_outputs.empty())
+        gpu_outputs.resize(num_samples);
+      if (!rois.empty())
+        rois.resize(num_samples);
+    }
+
+    /**
+     * @brief Moves one work entry from another work to this one
+     */
+    void move_entry(ScheduledWork &from, int which);
+
+    /**
+     * @brief Allocates temporary CPU outputs for this sub-batch
+     *
+     * This function is used when falling back from GPU to CPU decoder.
+     */
+    void alloc_temp_cpu_outputs(ImageDecoder &owner);
+
+    DecodeContext ctx;
+    // The original promise
+    DecodeResultsPromise results;
+    // The indices in the original request
+    BatchVector<int> indices;
+    BatchVector<ImageSource *> sources;
+    BatchVector<SampleView<CPUBackend>> cpu_outputs;
+    BatchVector<SampleView<GPUBackend>> gpu_outputs;
```

Comment:
how about having a single vector a struct with index, source, cpu_out, gpu_out...

Reply:
Nope, becasue we need spans of the separate entities to pass them to SchedulDecode.